File: src/main/java/com/github/qky666/selenidepom/RequiredContainer.java
package com.github.qky666.selenidepom;

import com.codeborne.selenide.CollectionCondition;
import com.codeborne.selenide.Condition;
import com.codeborne.selenide.Configuration;
import com.codeborne.selenide.ElementsCollection;
import com.codeborne.selenide.ElementsContainer;
import com.codeborne.selenide.Selenide;
import com.codeborne.selenide.SelenideElement;
import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

import javax.annotation.CheckReturnValue;
import javax.annotation.Nonnull;
import java.lang.reflect.Field;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.time.Duration;
import java.time.LocalDateTime;
import java.util.*;
import java.util.stream.Collectors;

/**
 * Object with fields and methods (without parameters) that can have @Required annotation.
 */
public interface RequiredContainer {
    /**
     * All fields and methods (without parameters) with @Required annotation are checked if visible.
     * You can override this method to add some extra functionality (custom additional checks).
     *
     * @param timeout The timeout for waiting to elements to become visible.
     * @throws RequiredError Error can occur during validations (mostly, validation failures).
     */
    default void shouldLoadRequiredWithTimeout(Duration timeout) throws RequiredError {
        final List<Throwable> errors = objectShouldLoadRequired(this, LocalDateTime.now().plus(timeout));
        if (!errors.isEmpty()) {
            throw new RequiredError(errors);
        }
    }

    /**
     * All fields and methods (without parameters) with @Required annotation are checked if visible, using the default timeout (taken from Selenide Configuration).
     * You usually override {@link #shouldLoadRequiredWithTimeout(Duration timeout) shouldLoadRequiredWithTimeout} instead, unless you need to change the default timeout.
     *
     * @throws RequiredError Error can occur during validations (mostly, validation failures).
     */
    default void shouldLoadRequired() throws RequiredError {
        shouldLoadRequiredWithTimeout(Duration.ofMillis(Configuration.timeout));
    }

    /**
     * Returns true if {@link #shouldLoadRequiredWithTimeout(Duration timeout) shouldLoadRequiredWithTimeout(timeout)} returns without throwing any exception, false otherwise.
     * You usually will not have to override this method.
     *
     * @param timeout The timeout for waiting to elements to become visible.
     * @return true if {@link #shouldLoadRequiredWithTimeout(Duration timeout) shouldLoadRequiredWithTimeout(timeout)} returns without throwing any exception, false otherwise.
     */
    @CheckReturnValue
    default boolean hasLoadedRequiredWithTimeout(Duration timeout) {
        return objectShouldLoadRequired(this, LocalDateTime.now().plus(timeout)).isEmpty();
    }

    /**
     * Returns true if {@link #shouldLoadRequiredWithTimeout(Duration timeout) shouldLoadRequiredWithTimeout(Duration.ZERO)} returns without throwing any exception, false otherwise.
     * You usually will not have to override this method.
     *
     * @return true if {@link #shouldLoadRequiredWithTimeout(Duration timeout) shouldLoadRequiredWithTimeout(Duration.ZERO)} returns without throwing any exception, false otherwise.
     */
    @CheckReturnValue
    default boolean hasLoadedRequired() {
        return hasLoadedRequiredWithTimeout(Duration.ZERO);
    }

    @CheckReturnValue
    @Nonnull
    static List<Throwable> objectShouldLoadRequired(Object object, LocalDateTime end) {
        List<Throwable> errors = new ArrayList<>();
        Set<String> processedFields = Collections.synchronizedSet(new HashSet<>());
        Set<String> processedMethods = Collections.synchronizedSet(new HashSet<>());

        Class<?> currentClass = object instanceof Class<?> ? (Class<?>) object : object.getClass();
        while (currentClass != Object.class) {
            // Fields
            List<Field> fields = Arrays.stream(currentClass.getDeclaredFields())
                    .filter(field -> {
                        field.setAccessible(true);
                        return processedFields.add(field.getName());
                    }).filter(field -> field.isAnnotationPresent(Required.class))
                    .collect(Collectors.toList());
            for (Field field : fields) {
                try {
                    Object element = field.get(object);
                    errors.addAll(elementShouldLoad(element, end));
                } catch (IllegalAccessException ignored) {
                }
            }

            // Methods
            List<Method> methods = Arrays.stream(currentClass.getDeclaredMethods())
                    .filter(method -> {
                        method.setAccessible(true);
                        return processedMethods.add(method.getName());
                    }).filter(method -> method.isAnnotationPresent(Required.class))
                    .collect(Collectors.toList());
            for (Method method : methods) {
                try {
                    Object element = method.invoke(object);
                    errors.addAll(elementShouldLoad(element, end));
                } catch (InvocationTargetException | IllegalAccessException ignored) {
                }
            }

            currentClass = currentClass.getSuperclass();
        }
        return errors;
    }

    @CheckReturnValue
    @Nonnull
    static List<Throwable> elementShouldLoad(Object element, LocalDateTime end) {
        Duration timeout = Duration.between(LocalDateTime.now(), end);
        if (timeout.isNegative()) {
            timeout = Duration.ZERO;
        }
        if (element instanceof By) {
            try {
                Selenide.$((By) element).shouldBe(Condition.visible, timeout);
            } catch (Throwable e) {
                return Collections.singletonList(e);
            }
        } else if (element instanceof SelenideElement) {
            try {
                ((SelenideElement) element).shouldBe(Condition.visible, timeout);
            } catch (Throwable e) {
                return Collections.singletonList(e);
            }
        } else if (element instanceof ElementsCollection) {
            try {
                ((ElementsCollection) element).shouldBe(CollectionCondition.anyMatch("At least one element is visible", WebElement::isDisplayed), timeout);
            } catch (Throwable e) {
                return Collections.singletonList(e);
            }
        } else if (element instanceof ElementsContainer) {
            try {
                ((ElementsContainer) element).getSelf().shouldBe(Condition.visible, timeout);
                return objectShouldLoadRequired(element, end);
            } catch (Throwable e) {
                return Collections.singletonList(e);
            }
        } else if (element instanceof RequiredContainer) {
            return objectShouldLoadRequired(element, end);
        } else if (element instanceof WebElement) {
            try {
                new WebDriverWait(Selenide.webdriver().object(), timeout).until(ExpectedConditions.visibilityOf((WebElement) element));
            } catch (Throwable e) {
                return Collections.singletonList(e);
            }
        } else {
            return objectShouldLoadRequired(element, end);
        }
        return Collections.emptyList();
    }
}

File: src/main/kotlin/com/github/qky666/selenidepom/kotlin/RequiredContainer.kt
package com.github.qky666.selenidepom.kotlin

import com.codeborne.selenide.*
import org.openqa.selenium.By
import org.openqa.selenium.WebElement
import org.openqa.selenium.support.ui.ExpectedConditions
import org.openqa.selenium.support.ui.WebDriverWait
import java.time.Duration
import java.time.LocalDateTime
import kotlin.reflect.full.hasAnnotation
import kotlin.reflect.full.memberProperties
import kotlin.reflect.jvm.isAccessible
import kotlin.reflect.jvm.javaField
import kotlin.reflect.jvm.javaGetter
import mu.KotlinLogging

private val logger = KotlinLogging.logger {}

/**
 * Instances of this interface can have properties that can be annotated with [Required].
 */
interface RequiredContainer {
    /**
     * All properties with [Required] annotation are checked if visible.
     * You can override this method to add some extra functionality (custom additional checks).
     *
     * @param timeout The timeout waiting for elements to become visible.
     * @throws RequiredError Error can occur during validations (mostly, validation failures).
     */
    @Throws(RequiredError::class)
    fun shouldLoadRequired(timeout: Duration = Duration.ofMillis(Configuration.timeout)) {
        val errors = objectShouldLoadRequired(this, LocalDateTime.now().plus(timeout))
        if (errors.isNotEmpty()) {
            throw RequiredError(errors)
        }
    }

    /**
     * Returns true if [shouldLoadRequired] returns without throwing any exception, false otherwise.
     * You usually will not have to override this method.
     *
     * @param timeout The timeout waiting for elements to become visible.
     * @return true if [shouldLoadRequired] returns without throwing any exception, false otherwise.
     */
    fun hasLoadedRequired(timeout: Duration = Duration.ZERO): Boolean {
        return objectShouldLoadRequired(this, LocalDateTime.now().plus(timeout)).isEmpty()
    }

    companion object {
        private fun objectShouldLoadRequired(obj: Any, end: LocalDateTime): List<Throwable> {
            val errors = mutableListOf<Throwable>()
            val objKlass = obj::class

            // Kotlin properties
            objKlass.memberProperties.forEach {
                it.isAccessible = true
                if (it.hasAnnotation<Required>()) {
                    val element = it.javaGetter?.invoke(obj) ?: it.javaField?.get(obj)
                    errors.addAll(elementShouldLoad(element, end))
                }
            }
            return errors
        }

        private fun elementShouldLoad(element: Any?, end: LocalDateTime): List<Throwable> {
            val signedTimeout: Duration = Duration.between(LocalDateTime.now(), end)
            val timeout: Duration = if (signedTimeout.isNegative) Duration.ZERO else signedTimeout
            when (element) {
                null -> return listOf()
                is By -> try {
                    Selenide.element(element).shouldBe(Condition.visible, timeout)
                    logger.info { "Checked element is visible (By): ${element.toString().replace("\n", "\\n")}" }
                } catch (e: Throwable) {
                    return listOf(e)
                }
                is SelenideElement -> try {
                    element.shouldBe(Condition.visible, timeout)
                    logger.info {
                        "Checked element is visible (SelenideElement): ${
                            element.toString().replace("\n", "\\n")
                        }"
                    }
                } catch (e: Throwable) {
                    return listOf(e)
                }
                is ElementsCollection -> try {
                    element.shouldBe(
                        CollectionCondition.anyMatch(
                            "At least one element is visible", WebElement::isDisplayed
                        ), timeout
                    )
                    logger.info {
                        "Checked at least one element is visible (ElementsCollection): ${
                            element.toString().replace("\n", "\\n")
                        }"
                    }
                } catch (e: Throwable) {
                    return listOf(e)
                }
                is ElementsContainerWidget -> return try {
                    element.self.shouldBe(Condition.visible, timeout)
                    logger.info {
                        "Checked element is visible (ElementsContainerWidget): ${
                            element.self.toString().replace("\n", "\\n")
                        }"
                    }
                    objectShouldLoadRequired(element, end)
                } catch (e: Throwable) {
                    listOf(e)
                }
                is ElementsContainer -> return try {
                    element.self.shouldBe(Condition.visible, timeout)
                    logger.info {
                        "Checked element is visible (ElementsContainer): ${
                            element.self.toString().replace("\n", "\\n")
                        }"
                    }
                    objectShouldLoadRequired(element, end)
                } catch (e: Throwable) {
                    listOf(e)
                }
                is Widget -> return try {
                    element.self.shouldBe(Condition.visible, timeout)
                    logger.info {
                        "Checked element is visible (Widget): ${
                            element.self.toString().replace("\n", "\\n")
                        }"
                    }
                    objectShouldLoadRequired(element, end)
                } catch (e: Throwable) {
                    listOf(e)
                }
                is RequiredContainer -> return objectShouldLoadRequired(element, end)
                is WebElement -> try {
                    WebDriverWait(Selenide.webdriver().`object`(), timeout).until(
                        ExpectedConditions.visibilityOf(element)
                    )
                    logger.info { "Checked element is visible (WebElement): ${element.toString().replace("\n", "\\n")}" }
                } catch (e: Throwable) {
                    return listOf(e)
                }
                else -> return objectShouldLoadRequired(element, end)
            }
            return listOf()
        }
    }
}